cms/djangoapps/contentstore/tasks.py

```diff
@@ -125,3 +153,218 @@ def push_course_update_task(course_key_string, course_subscription_id, course_di
     # TODO Use edx-notifications library instead (MA-638).
     from .push_notification import send_push_course_update
     send_push_course_update(course_key_string, course_subscription_id, course_display_name)
+
+
+class CourseImportTask(UserTask):  # pylint: disable=abstract-method
+    """
+    Base class for course and library import tasks.
+    """
+
+    @staticmethod
+    def calculate_total_steps(arguments_dict):
+        """
+        Get the number of in-progress steps in the import process, as shown in the UI.
+
+        For reference, these are:
+
+        1. Unpacking
+        2. Verifying
+        3. Updating
+        """
+        return 3
+
+    @classmethod
+    def generate_name(cls, arguments_dict):
+        """
+        Create a name for this particular import task instance.
+
+        Arguments:
+            arguments_dict (dict): The arguments given to the task function
+
+        Returns:
+            text_type: The generated name
+        """
+        key = arguments_dict[u'course_key_string']
+        filename = arguments_dict[u'archive_name']
+        return u'Import of {} from {}'.format(key, filename)
+
+
+@task(base=CourseImportTask, bind=True)
+def import_olx(self, user_id, course_key_string, archive_path, archive_name, language):
+    """
+    Import a course or library from a provided OLX .tar.gz archive.
+    """
+    courselike_key = CourseKey.from_string(course_key_string)
+    try:
+        user = User.objects.get(pk=user_id)
+    except User.DoesNotExist:
+        with respect_language(language):
+            self.status.fail(_(u'Unknown User ID: {0}').format(user_id))
+        return
+    if not has_course_author_access(user, courselike_key):
+        with respect_language(language):
+            self.status.fail(_(u'Permission denied'))
+        return
+
+    is_library = isinstance(courselike_key, LibraryLocator)
+    is_course = not is_library
+    if is_library:
+        root_name = LIBRARY_ROOT
+        courselike_module = modulestore().get_library(courselike_key)
+        import_func = import_library_from_xml
+    else:
+        root_name = COURSE_ROOT
+        courselike_module = modulestore().get_course(courselike_key)
+        import_func = import_course_from_xml
+
+    # Locate the uploaded OLX archive (and download it from S3 if necessary)
+    # Do everything in a try-except block to make sure everything is properly cleaned up.
+    data_root = path(settings.GITHUB_REPO_ROOT)
+    subdir = base64.urlsafe_b64encode(repr(courselike_key))
+    course_dir = data_root / subdir
+    try:
+        self.status.set_state(u'Unpacking')
+
+        if not archive_name.endswith(u'.tar.gz'):
+            with respect_language(language):
+                self.status.fail(_(u'We only support uploading a .tar.gz file.'))
```

Comment:
I see - you're referring to the *previous* upload? The one that got the `.tar.gz` file there in the first place? Since the download of this file happens below, isn't this code refusing to download anything that's not a `.tar.gz` file?